### ISSUE TITLE
[FR] hub uses default github branch

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#43638]
Currently hub uses the master branch always. But on repos with a custom default branch, wouldn't it be nicer if we use that?

**Comment by ezyang** [comment]
I agree, this seems like a reasonable thing to do.

**Comment by SsnL** [comment]
Code-wise, this amounts to making an API call (don't need token) like
```
curl \
  -H "Accept: application/vnd.github.v3+json" \
  https://api.github.com/repos/{user}/{repo}
```
and read the `default_branch` field.

**Comment by ezyang** [comment]
Doesn't GitHub say default branch info shows up in ls-remote too? Because I think when you do a git clone, you get the default branch.

**Comment by ezyang** [comment]
https://git-scm.com/docs/git-ls-remote.html

**Comment by VitalyFedyunin** [comment]
Not labeled properly.

**Comment by SsnL** [comment]
Yeah, but I think hub doesn't use that, and directly composes a download link instead using the master branch by default
https://github.com/pytorch/pytorch/blob/95a69a7d09e4089dad92e65f5d48ceeec70ba3b4/torch/hub.py#L103-L119
https://github.com/pytorch/pytorch/blob/95a69a7d09e4089dad92e65f5d48ceeec70ba3b4/torch/hub.py#L84-L85

**Comment by ailzhang** [comment]
Yea this makes sense! @SsnL Please feel free to send a PR to update it! (Or I can do it later this week :D
